batcher/sweep_batch.go

```diff
@@ -0,0 +1,1145 @@
+		PkScript: batchPkScript,
+		Value:    int64(batchAmt - fee),
+	})
+
+	// Collect the signatures for our inputs.
+	rawSigs, err := b.signerClient.SignOutputRaw(
+		ctx, batchTx, signDescs, prevOuts,
+	)
+	if err != nil {
+		return fee, err
+	}
+
+	for i, sweep := range sweeps {
+
+		// Generate the success witness for the sweep.
+		witness, err := sweep.Htlc.GenSuccessWitness(
+			rawSigs[i], sweep.Preimage,
+		)
+		if err != nil {
+			return fee, err
+		}
+
+		// Add the success witness to our batch transaction's inputs.
+		batchTx.TxIn[i].Witness = witness
+	}
+
+	log.Debugf("Batch %x: attempting to publish tx with feerate=%v, "+
+		"totalfee=%v", b.ID[:6], b.RbfCache.FeeRate, fee)
+	err = b.wallet.PublishTransaction(
+		ctx, batchTx, labels.LoopOutBatchSweepSuccess(b.ID.String()),
+	)
+	if err != nil {
+		return fee, err
+	}
+
+	// Store the batch transaction's txid and pkScript, for monitoring
+	// purposes.
+	txHash := batchTx.TxHash()
+	b.BatchTxid = &txHash
+	b.BatchPkScript = batchPkScript
+
+	return fee, nil
+}
+
+// publishBatchCoop attempts to construct and publish a batch transaction that
+// collects all the required signatures interactively from the server. This
+// helps with collecting the funds immediately without revealing any information
+// related to the HTLC script.
+func (b *Batch) publishBatchCoop(ctx context.Context) (btcutil.Amount,
+	error, bool) {
+	batchAmt := btcutil.Amount(0)
+	sweeps := make([]Sweep, 0, len(b.Sweeps))
+	fee := btcutil.Amount(0)
+	var weightEstimate input.TxWeightEstimator
+
+	// Create the batch transaction.
+	batchTx := &wire.MsgTx{
+		Version:  2,
+		LockTime: uint32(b.currentHeight),
+	}
+
+	for _, sweep := range b.Sweeps {
+		// Append this sweep to an array of sweeps. This is needed to
+		// keep the order of sweeps stored, as iterating the sweeps map
+		// does not guarantee same order.
+		sweeps = append(sweeps, sweep)
+	}
+
+	// Add all the sweeps to the batch transaction.
+	for _, sweep := range sweeps {
+		// Keep track of the total amount this batch is sweeping back.
+		batchAmt += sweep.Value
+
+		// Add this sweep's input to the transaction.
+		batchTx.AddTxIn(&wire.TxIn{
+			PreviousOutPoint: sweep.Outpoint,
+		})
+
+		updateWeightEstimate(&weightEstimate, sweep.Htlc.PkScript, true)
+	}
+
+	// Generate a wallet address for the batch transaction's output.
+	address, err := b.wallet.NextAddr(
+		ctx, "", walletrpc.AddressType_TAPROOT_PUBKEY, false,
+	)
+	if err != nil {
+		return fee, err, false
+	}
+
+	batchPkScript, err := txscript.PayToAddrScript(address)
+	if err != nil {
+		return fee, err, false
+	}
+
+	weightEstimate.AddP2TROutput()
+
+	totalWeight := int64(weightEstimate.Weight())
+
+	fee = b.RbfCache.FeeRate.FeeForWeight(totalWeight)
+
+	// Add the batch transaction output, which excludes the fees paid to
+	// miners.
+	batchTx.AddTxOut(&wire.TxOut{
+		PkScript: batchPkScript,
+		Value:    int64(batchAmt - fee),
+	})
+
+	packet, err := psbt.NewFromUnsignedTx(batchTx)
```

Comment:
I think we'd also need to sort the inputs and outputs to conform to BIP-69. AFAICT psbt lib doesn't do that.
```
txsort.InPlaceSort(batchTx)
```